### PR TITLE
RFC Firmware downgrading [WiFi hotspot capacity workaround, for Raspberry Pi]

### PR DESCRIPTION
**Pull request by jvonau**
only forces downgrade with no provisions to undo the rolling back of the firmware.

### CODE REVIEW

File: roles/firmware/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  include_tasks: install.yml
+  when: RPi_firmware_retrieved is undefined
+
+- name: Install check-firmware service files
+  template:
+    src: "{{ item.src }}"
+    dest: "{{ item.dest }}"
+    mode: "{{ item.mode }}"
+  with_items:
+    - { src: 'fw-warn.sh', dest: '/usr/sbin/' mode: '0644' }
```
Comment (holta) [inline roles/firmware/tasks/main.yml]: fw-warn.sh is a better name than fw_warn.sh below, if this is needed.
```suggestion
    - { src: 'fw-warn.sh', dest: '/usr/sbin/', mode: '0644' }
```

File: roles/firmware/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    - { src: 'check-firmware.service', dest: '/etc/systemd/system/' mode: '0644'}
```
Comment (holta) [inline roles/firmware/tasks/main.yml]:
```suggestion
    - { src: 'check-firmware.service', dest: '/etc/systemd/system/', mode: '0644' }
```

File: roles/firmware/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    - { src: 'check-firmware.sh', dest: '/usr/sbin/' mode: '0755' }
```
Comment (holta) [inline roles/firmware/tasks/main.yml]:
```suggestion
    - { src: 'check-firmware.sh', dest: '/usr/sbin/', mode: '0755' }
```

**Review by holta — COMMENTED** [review]
Just a quick review (inlined comments) noticing a few small bugs.

Separately, I personally I lean towards:

1.  Including a backup every time firmware file(s) are overwritten (so the operator is protected no matter what, always able to roll back firmware if truly needed).  Similar to suggestions at: https://github.com/iiab/iiab/issues/823#issuecomment-667563551
2. Preserving the timestamp whenever firmware files are copied or backed up, making all these files (and a bit of their history) easier to understand, e.g. using `cp -p` or `rsync -t`

Tangentially Related:

PR #2471 "RPiOS - use dhcpcd hooks to restart hostapd when br0 appears [for Ubuntu too]"

File: roles/firmware/templates/fw_warn.sh [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (jvonau) [reply roles/firmware/templates/fw_warn.sh:1]: latest commit

**Review by holta — COMMENTED** [review]
Two (in the end three) missing commas.

**Comment by jvonau** [comment]
There are 3 diffs being testing the 3 files, remember the stock 43430 files doesn't provide a clm_blob, only the replacement bin requires it. What is your issue?

**Comment by holta** [comment]
1) I see that 3rd test now, understood.

2) If either .clm_blob changes, WiFi (generally) becomes very unreliable, in a classroom situation of many WiFi devices especially, as my July tests confirmed.

So testing for future changes to both .clm_blob files is probably wise (adding a 4th test).

That 4th file being http://d.iiab.io/packages/brcmfmac43455-sdio.clm_blob_2018-02-26_rpi (currently shipping with RaspiOS & Ubuntu-for-RPi both, but that could change in future).

**Comment by holta** [comment]
While testing this, I see "RPi_firmware_retrieved: True" in /etc/iiab/iiab_state.yml

If possible can we keep this lowercase, and matching the exact name of the Ansible role, to help everyone keep track going forward?  (Possibly the role should be renamed to rpi_firmware ?)

(In any case keeping these names self-consistent wherever they appear, helping avoid many ambiguities, e.g. when running ./runrole etc.)

**Comment by holta** [comment]
fyi this PR #2472 (and PR #2471) are being tested together with PR #2474, on an RPi 4 @ 10.8.0.14

**Comment by holta** [comment]
Tested on RPi 4 2GB running RaspiOS Lite.

**Pull request merged**